networkx/algorithms/shortest_paths/unweighted.py

```diff
@@ -29,12 +29,14 @@ def single_source_shortest_path_length(G, source, cutoff=None):
        Starting node for path
 
     cutoff : integer, optional
-        Depth to stop the search. Only paths of length <= cutoff are returned.
+        Depth to stop the search. Only target nodes where the shortest path to
+        this node from the source node contains <= `cutoff` edges will be
+        included in the returned results.
```

Comment:
I don't think the proposed wording is necessarily an improvement. For one thing, the verbosity makes it harder to grok (though that's the price for greater precision).

The bigger issue IMO is the fact that introducing "edges" to the explanation is not strictly correct... For example, in the case of multigraphs, it's the existence of an edge between nodes (rather than a specific edge) that dictates whether there is a path between them. Therefore, the number of edges is ambiguous in this context.

I have an alternative proposal: rather than modify the wording in the parameter description, what would you think about instead adding a short example to the `Examples` section of the docstrings that illustrates how `cutoff` works? e.g. something like (for this function):

```python
>>> G = nx.path_graph(3)
>>> nx.single_source_shortest_path_length(G, 2)
{0: 0, 1: 1, 2: 2}
>>> nx.single_source_shortest_path_length(G, 2, cutoff=1)
{0: 0, 1: 1}
```

Reply:
Thank you for the comments!

> I don't think the proposed wording is necessarily an improvement. For one thing, the verbosity makes it harder to grok (though that's the price for greater precision).

I wouldn't mind shortening the proposed wording, but I believe we should at least include the word "edges" in these descriptions. When I was reading the docs for `all_simple_paths()`, I assumed that path length was given in terms of nodes, not edges, and I imagine other people will make similar mistakes if these descriptions are left ambiguous.

For this particular function (`single_source_shortest_path_length()`) the meaning of "path length" is a bit clearer than for `all_simple_paths()` (since the output of the function actually contains path lengths, unlike the output of `all_simple_paths()`), but I still think more precision about the meaning of "path length" would be helpful in both cases.

> The bigger issue IMO is the fact that introducing "edges" to the explanation is not strictly correct... For example, in the case of multigraphs, it's the existence of an edge between nodes (rather than a specific edge) that dictates whether there is a path between them. Therefore, the number of edges is ambiguous in this context.

I apologize if I have misunderstood something, but I do not agree with this logic. Consider the following multidigraph:

![image](https://github.com/user-attachments/assets/4d81e79f-b2b7-4d24-8ed8-439ccea07dbd)

Although there are four edges in this graph, and although any of these four edges could be included on a shortest path from 0 to 2, the lengths of these shortest paths (in terms of edges, per `single_source_shortest_path_length()`) are all 2. The function on which we are currently commenting agrees with this:

```python
import networkx as nx
g = nx.MultiDiGraph()
g.add_edge(0, 1, key="A")
g.add_edge(0, 1, key="B")
g.add_edge(0, 1, key="C")
g.add_edge(1, 2, key="D")
assert nx.single_source_shortest_path_length(g, 0) == {0: 0, 1: 1, 2: 2}
```

I do not think it is conventional to define the number of edges on a path relative to the number of "possible" edges on this path -- instead, this would be expressed as the number of such paths. In this example:

- `the shortest path to 0 from 0` contains 0 edges,
- `the shortest path to 1 from 0` contains 1 edge, and
- `the shortest path to 2 from 0` contains 2 edges,

even though there are 3 possible shortest paths from 0 to 1 and from 0 to 2.

As far as I am aware, the proposed wordings in this PR are correct, but please let me know if I've missed something.

> I have an alternative proposal: rather than modify the wording in the parameter description, what would you think about instead adding a short example to the Examples section of the docstrings that illustrates how cutoff works?

Sure! Of the eight functions whose docstrings were updated in this PR, only one of them (`all_simple_paths()`) contains an example that uses the `cutoff` parameter. I just added examples for the other seven functions to this PR.

Although I hope that these examples will be helpful, I do not think that users should be required to read through them to understand how exactly a parameter works -- particularly when the ambiguity about path length discussed above can be fixed by adding/adjusting a few words. So, my preference is to both (1) keep these additional examples and (2) keep the proposed clarifications that "path length" is given in terms of edges. I would be happy to pare down the proposed wordings if you feel that they are currently too difficult to read.

Reply:
> In other words, IMV it's more "networkx-y" to consider paths from the perspective of a sequence of nodes rather than making explicit mention to edges.

I agree with this sentiment, but unfortunately all of these functions already implicitly define "path length" to be in terms of edges :)

> FWIW, NetworkX generally treats paths (in the context of the shortest/simple path finding functions, anyways) as a sequence of nodes.

I suspect that what you are describing here is part of the reason _why_ the current descriptions of `cutoff` are prone to misinterpretation. Since [`all_simple_paths()`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.simple_paths.all_simple_paths.html) (and [`single_source_shortest_path()`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.unweighted.single_source_shortest_path.html), etc.) return paths as lists of nodes, it is _surprising_ to users that these functions' `cutoff` parameters are in terms of edges, not nodes.

As an example -- I would guess that many people (if given just the current main branch's [`single_source_shortest_path()`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.unweighted.single_source_shortest_path.html) documentation, and no ability to e.g. try this function out in a REPL) would assume that `cutoff` would be in terms of nodes, since nothing in this function's documentation indicates anything to the contrary. **To prevent users from writing subtly buggy code, I believe we should be very clear that this is not the case.** Adding more examples (as done in this PR, now) will help, but this is not a substitute for explicit documentation.

It sounds like we agree that the ideal solution would be going back in time and changing all of these functions' `cutoff` parameters to be in terms of nodes instead of edges -- unfortunately, I don't think that is feasible any more (at the very least, such a change would probably be out of scope of this PR). For now, my opinion (on this admittedly subjective situation) is that we should be clear about this, and admit that `cutoff` is in terms of edges instead of nodes.